src/clients/HubPoolClient.ts

```diff
@@ -165,74 +165,91 @@ export class HubPoolClient extends BaseAbstractClient {
+    if (!this.l1TokensToDestinationTokensWithBlock?.[l1Token]?.[destinationChainId]) {
+      throw new Error(`Could not find L2 token mapping for chain ${destinationChainId} and L1 token ${l1Token}`);
     }
-    return destinationToken;
-  }
-
-  getL1TokensToDestinationTokens(): L1TokensToDestinationTokens {
-    return this.l1TokensToDestinationTokens;
-  }
-
-  getL1TokenForDeposit(deposit: { originChainId: number; originToken: string }): string {
-    const l1Token = Object.keys(this.l1TokensToDestinationTokens).find((_l1Token) => {
-      return this.l1TokensToDestinationTokens[_l1Token][deposit.originChainId] === deposit.originToken;
-    });
-    if (!l1Token) {
+    // Find the last mapping published before the target block.
+    const l2Token: DestinationTokenWithBlock | undefined = sortEventsDescending(
+      this.l1TokensToDestinationTokensWithBlock[l1Token][destinationChainId]
+    ).find((mapping: DestinationTokenWithBlock) => mapping.blockNumber <= latestHubBlock);
+    if (!l2Token) {
       throw new Error(
-        `Could not find L1 Token for origin chain ${deposit.originChainId} and origin token ${deposit.originToken}!`
+        `Could not find L2 token mapping for chain ${destinationChainId} and L1 token ${l1Token} equal to or earlier than block ${latestHubBlock}!`
       );
     }
-    return l1Token;
-  }
-
-  getL1TokenCounterpartAtBlock(l2ChainId: number, l2Token: string, hubPoolBlock: number): string {
-    const l1Token = Object.keys(this.l1TokensToDestinationTokensWithBlock).find((_l1Token) => {
-      // If this token doesn't exist on this L2, return false.
-      if (this.l1TokensToDestinationTokensWithBlock[_l1Token][l2ChainId] === undefined) {
-        return false;
-      }
-
-      // Find the last mapping published before the target block.
-      return sortEventsDescending(this.l1TokensToDestinationTokensWithBlock[_l1Token][l2ChainId]).find(
-        (mapping: DestinationTokenWithBlock) => mapping.l2Token === l2Token && mapping.blockNumber <= hubPoolBlock
-      );
-    });
-    if (!l1Token) {
+    return l2Token.l2Token;
+  }
+
+  // Returns the latest L1 token to use for an L2 token as of the input hub block.
+  protected getL1TokenForL2TokenAtBlock(
+    l2Token: string,
+    destinationChainId: number,
+    latestHubBlock = Number.MAX_SAFE_INTEGER
+  ): string {
+    const l2Tokens = Object.keys(this.l1TokensToDestinationTokensWithBlock)
+      .map((l1Token) => {
+        // If this token doesn't exist on this L2, skip it
+        if (this.l1TokensToDestinationTokensWithBlock[l1Token][destinationChainId] === undefined) {
+          return undefined;
+        }
+        // Return all matching L2 token mappings that are equal to or earlier than the target block.
+        return this.l1TokensToDestinationTokensWithBlock[l1Token][destinationChainId].filter(
+          (mapping) => mapping.l2Token === l2Token && mapping.blockNumber <= latestHubBlock
+        );
+      })
+      .filter(isDefined)
```

Comment:
Seems like we can use `this.l2TokenEnabledForL1Token()` here?

```suggestion
      .filter((l1Token) => this.l2TokenEnabledForL1Token(l1Token, destinationChainId)
      .map((l1Token) => {
        // Return all matching L2 token mappings that are equal to or earlier than the target block.
        return this.l1TokensToDestinationTokensWithBlock[l1Token][destinationChainId].filter(
          (mapping) => mapping.l2Token === l2Token && mapping.blockNumber <= latestHubBlock
        );
      });
```